<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · CodonMolecularEvolution.jl</title><script data-outdated-warner src="assets/warner.js"></script><link rel="canonical" href="https://MurrellGroup.github.io/CodonMolecularEvolution.jl/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script><link href="assets/favicon.ico" rel="icon" type="image/x-icon"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href><img src="assets/logo.png" alt="CodonMolecularEvolution.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href>CodonMolecularEvolution.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li><li><a class="tocitem" href="difFUBAR/">difFUBAR</a></li><li><a class="tocitem" href="shared/">Shared tools</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/main/docs/src/index.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="CodonMolecularEvolution"><a class="docs-heading-anchor" href="#CodonMolecularEvolution">CodonMolecularEvolution</a><a id="CodonMolecularEvolution-1"></a><a class="docs-heading-anchor-permalink" href="#CodonMolecularEvolution" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl">CodonMolecularEvolution</a>.</p><p>Descendant of <a href="https://github.com/MurrellGroup/MolecularEvolution.jl">MolecularEvolution.jl</a>, specializing in codon models.</p><h3 id="A-Julia-package-that-contains-a-collection-of-popular/(state-of-the-art)-codon-model-methods."><a class="docs-heading-anchor" href="#A-Julia-package-that-contains-a-collection-of-popular/(state-of-the-art)-codon-model-methods.">A Julia package that contains a collection of popular/(state of the art) codon model methods.</a><a id="A-Julia-package-that-contains-a-collection-of-popular/(state-of-the-art)-codon-model-methods.-1"></a><a class="docs-heading-anchor-permalink" href="#A-Julia-package-that-contains-a-collection-of-popular/(state-of-the-art)-codon-model-methods." title="Permalink"></a></h3><p>CodonMolecularEvolution.jl leverages the MolecularEvolution.jl framework to provide a performant interface to some codon model methods that are used for phylogenetic analysis.</p><h3 id="Collection-of-codon-model-methods"><a class="docs-heading-anchor" href="#Collection-of-codon-model-methods">Collection of codon model methods</a><a id="Collection-of-codon-model-methods-1"></a><a class="docs-heading-anchor-permalink" href="#Collection-of-codon-model-methods" title="Permalink"></a></h3><ul><li><a href="difFUBAR/#difFUBAR">difFUBAR</a>: Scalable Bayesian comparison of selection pressure<ul><li>Perform a site-wise comparison of evolutionary pressure between two selected sets of branches.</li></ul></li></ul><h3 id="Design-principles"><a class="docs-heading-anchor" href="#Design-principles">Design principles</a><a id="Design-principles-1"></a><a class="docs-heading-anchor-permalink" href="#Design-principles" title="Permalink"></a></h3><ul><li>Specificity<ul><li>We keep our implementations on a high level of abstraction and let <code>MolecularEvolution.jl</code> take care of the low-level behaviour.</li></ul></li><li>Scalability<ul><li>Our exported codon model methods should scale to large, real-world datasets. To keep the memory footprint down, we use <code>MolecularEvolution.jl</code>s <code>LazyPartition</code> when possible.</li></ul></li><li>Performance<ul><li>While scalability takes precedence over speed?, we try to maintain competitive runtimes by using e.g. computational shortcuts and parallelization whenever we can.</li></ul></li></ul><h3 id="Authors"><a class="docs-heading-anchor" href="#Authors">Authors</a><a id="Authors-1"></a><a class="docs-heading-anchor-permalink" href="#Authors" title="Permalink"></a></h3><p>Ben decides what to put here.</p><ul><li><a href="difFUBAR/#CodonMolecularEvolution.difFUBARBaseline"><code>CodonMolecularEvolution.difFUBARBaseline</code></a></li><li><a href="difFUBAR/#CodonMolecularEvolution.difFUBARParallel"><code>CodonMolecularEvolution.difFUBARParallel</code></a></li><li><a href="difFUBAR/#CodonMolecularEvolution.difFUBARTreesurgery"><code>CodonMolecularEvolution.difFUBARTreesurgery</code></a></li><li><a href="difFUBAR/#CodonMolecularEvolution.difFUBARTreesurgeryAndParallel"><code>CodonMolecularEvolution.difFUBARTreesurgeryAndParallel</code></a></li><li><a href="#CodonMolecularEvolution.benchmark_global_fit-Tuple{Any}"><code>CodonMolecularEvolution.benchmark_global_fit</code></a></li><li><a href="#CodonMolecularEvolution.benchmark_grid-Tuple{Any}"><code>CodonMolecularEvolution.benchmark_grid</code></a></li><li><a href="#CodonMolecularEvolution.difFUBAR-NTuple{6, Any}"><code>CodonMolecularEvolution.difFUBAR</code></a></li><li><a href="difFUBAR/#CodonMolecularEvolution.difFUBAR"><code>CodonMolecularEvolution.difFUBAR</code></a></li><li><a href="#CodonMolecularEvolution.difFUBAR_init-NTuple{4, Any}"><code>CodonMolecularEvolution.difFUBAR_init</code></a></li><li><a href="#CodonMolecularEvolution.getpuresubclades-Tuple{MolecularEvolution.FelNode, Vector{String}}"><code>CodonMolecularEvolution.getpuresubclades</code></a></li><li><a href="#CodonMolecularEvolution.import_colored_figtree_nexus_as_tagged_tree-Tuple{Any}"><code>CodonMolecularEvolution.import_colored_figtree_nexus_as_tagged_tree</code></a></li><li><a href="shared/#CodonMolecularEvolution.import_colored_figtree_nexus_as_tagged_tree"><code>CodonMolecularEvolution.import_colored_figtree_nexus_as_tagged_tree</code></a></li><li><a href="#CodonMolecularEvolution.import_grouped_label_tree-Tuple{Any}"><code>CodonMolecularEvolution.import_grouped_label_tree</code></a></li><li><a href="shared/#CodonMolecularEvolution.import_grouped_label_tree"><code>CodonMolecularEvolution.import_grouped_label_tree</code></a></li><li><a href="#CodonMolecularEvolution.optimize_MG94_F3x4-Tuple{Any, Any, Any}"><code>CodonMolecularEvolution.optimize_MG94_F3x4</code></a></li><li><a href="shared/#CodonMolecularEvolution.optimize_MG94_F3x4"><code>CodonMolecularEvolution.optimize_MG94_F3x4</code></a></li></ul><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.benchmark_global_fit-Tuple{Any}" href="#CodonMolecularEvolution.benchmark_global_fit-Tuple{Any}"><code>CodonMolecularEvolution.benchmark_global_fit</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">CodonMolecularEvolution.benchmark_global_fit(benchmark_name; exports=true, data=1:5, optimize_branch_lengths=false)</code></pre><p>Benchmarks different implementations of the difFUBAR<em>global</em>fit algorithm. Results of the benchmark are printed out as a DataFrame and saved to a CSV file.  Uses the heuristic top pick to generate con lik matrices. Compares difference in con lik matrices. If exports is true, it also runs MCMCs on the con lik matrices and plots the means and posteriors of the different versions.</p><ul><li><code>benchmark_name</code> is the filepath to where the benchmark will be saved, if exports.</li><li><code>data</code> is the range/vector of datasets to run the benchmark on. By default, this is 1:5. These are the enumerated datasets:<ul><li><ol><li>Ace2nobackground</li></ol></li><li><ol><li>Ace2reallytiny</li></ol></li><li><ol><li>Ace2tiny</li></ol></li><li><ol><li>ParvoVP</li></ol></li><li><ol><li>ParvoVPregrouped</li></ol></li></ul></li><li><code>optimize_branch_lengths</code> is an option that can be either <code>true</code>, <code>false</code> or <code>&quot;detect&quot;</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ab9a6cbb74e0df49dd0ed6151cc023e8eb98181f/test/benchmark_difFUBAR.jl#L253-L266">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.benchmark_grid-Tuple{Any}" href="#CodonMolecularEvolution.benchmark_grid-Tuple{Any}"><code>CodonMolecularEvolution.benchmark_grid</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">CodonMolecularEvolution.benchmark_grid(benchmark_name; exports=true, versions_option=1, t::Integer=0, data=1:5)</code></pre><p>Benchmarks different implementations of the difFUBAR_grid algorithm. Results of the benchmark are printed out as a DataFrame and saved to a CSV file.</p><ul><li><code>benchmark_name</code> is the filepath to where the benchmark will be saved, if exports</li><li><code>versions_option</code> have 4 different options:<ul><li><ol><li>default option, only run heuristic top pick</li></ol></li><li><ol><li>only run baseline version</li></ol></li><li><ol><li>run heuristic top pick and baseline version</li></ol></li><li><ol><li>run all versions</li></ol></li></ul></li><li><code>t</code> is the number of threads you want to use in the parallel versions. If specified and non-zero, this will override the number of threads chosen by the heuristic.</li><li><code>data</code> is the range/vector of datasets to run the benchmark on. By default, this is 1:5. These are the enumerated datasets:<ul><li><ol><li>Ace2nobackground</li></ol></li><li><ol><li>Ace2reallytiny</li></ol></li><li><ol><li>Ace2tiny</li></ol></li><li><ol><li>ParvoVP</li></ol></li><li><ol><li>ParvoVPregrouped</li></ol></li></ul></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ab9a6cbb74e0df49dd0ed6151cc023e8eb98181f/test/benchmark_difFUBAR.jl#L193-L209">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.difFUBAR-NTuple{6, Any}" href="#CodonMolecularEvolution.difFUBAR-NTuple{6, Any}"><code>CodonMolecularEvolution.difFUBAR</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">difFUBAR(seqnames, seqs, treestring, tags, tag_colors, outpath; &lt;keyword arguments&gt;)</code></pre><p>Takes a tagged phylogeny and an alignment as input and performs difFUBAR analysis.</p><p><strong>Arguments</strong></p><ul><li><code>seqnames</code>: vector of untagged sequence names.</li><li><code>seqs</code>: vector of aligned sequences, corresponding to <code>seqnames</code>.</li><li><code>treestring</code>: a tagged newick tree string.</li><li><code>tags</code>: vector of tag signatures.</li><li><code>tag_colors</code>: vector of tag colors (hex format).</li><li><code>outpath</code>: export directory.</li><li><code>pos_thresh=0.95</code>: threshold of significance for the posteriors.</li><li><code>iters=2500</code>: iterations used in the Gibbs sampler.</li><li><code>verbosity=1</code>: as verbosity increases, prints are added accumulatively. <ul><li>0 - no prints</li><li>1 - show current step and where output files are exported</li><li>2 - show the chosen <code>difFUBAR_grid</code> version and amount of parallel threads.</li></ul></li><li><code>exports=true</code>: if true, output files are exported.</li><li><code>code=MolecularEvolution.universal_code</code>: genetic code used for the analysis.</li><li><code>optimize_branch_lengths=false</code>: if true, the branch lengths of the phylogenetic tree are optimized.</li><li><code>version::Union{difFUBARGrid, Nothing}=nothing</code>: explicitly choose the version of <code>difFUBAR_grid</code> to use. If <code>nothing</code>, the version is heuristically chosen based on the available RAM and Julia threads.</li><li><code>t=0</code>: explicitly choose the amount of Julia threads to use. If <code>0</code>, the degree of parallelization is heuristically chosen based on the available RAM and Julia threads.</li></ul><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Julia starts up with a single thread of execution, by default. See <a href="https://docs.julialang.org/en/v1/manual/multi-threading/#Starting-Julia-with-multiple-threads">Starting Julia with multiple threads</a>.</p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ab9a6cbb74e0df49dd0ed6151cc023e8eb98181f/src/difFUBAR/difFUBAR.jl#L414-L440">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.difFUBAR_init-NTuple{4, Any}" href="#CodonMolecularEvolution.difFUBAR_init-NTuple{4, Any}"><code>CodonMolecularEvolution.difFUBAR_init</code></a> — <span class="docstring-category">Method</span></header><section><div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ab9a6cbb74e0df49dd0ed6151cc023e8eb98181f/src/difFUBAR/difFUBAR.jl#L137-L140">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.getpuresubclades-Tuple{MolecularEvolution.FelNode, Vector{String}}" href="#CodonMolecularEvolution.getpuresubclades-Tuple{MolecularEvolution.FelNode, Vector{String}}"><code>CodonMolecularEvolution.getpuresubclades</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">getpuresubclades(tree::FelNode, tags::Vector{String})</code></pre><ul><li>Should usually be called on the root of the tree. Traverses the tree iteratively with a depth-first search to find roots of pure subclades, presuming that nodenames have been trailed with tags. Returns a Vector{FelNode} with root-nodes of the pure subclades.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ab9a6cbb74e0df49dd0ed6151cc023e8eb98181f/src/difFUBAR/grids.jl#L92-L96">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.import_colored_figtree_nexus_as_tagged_tree-Tuple{Any}" href="#CodonMolecularEvolution.import_colored_figtree_nexus_as_tagged_tree-Tuple{Any}"><code>CodonMolecularEvolution.import_colored_figtree_nexus_as_tagged_tree</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">import_colored_figtree_nexus_as_tagged_tree(fname; custom_labels=String[])</code></pre><p>Takes a nexus file from FigTree, where branches have been colored. Replaces all color tags with group tags that can be used in the models. Can add custom labels too. Should consider an entire custom dictionary as well in future.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; treestring, tags, tag_colors = import_colored_figtree_nexus_as_tagged_tree(&quot;data/Ace2_no_background.nex&quot;)
(&quot;(((XM_027533928_Bos_indicus_x_Bos_taurus{G1}:0.097072,(XM_042974087_Panthera_tigris{G1}:0.038016,... more ...;&quot;, [&quot;{G2}&quot;, &quot;{G1}&quot;], [&quot;#ff0015&quot;, &quot;#0011ff&quot;])</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>treestring</code> is truncated. <a href="https://raw.githubusercontent.com/MurrellGroup/CodonMolecularEvolution.jl/main/test/data/Ace2_no_background/Ace2_no_background.nex">NEXUS tree file</a></p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ab9a6cbb74e0df49dd0ed6151cc023e8eb98181f/src/shared/shared.jl#L14-L26">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.import_grouped_label_tree-Tuple{Any}" href="#CodonMolecularEvolution.import_grouped_label_tree-Tuple{Any}"><code>CodonMolecularEvolution.import_grouped_label_tree</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">import_grouped_label_tree(tree_file)</code></pre><p>Takes a Newick tree file and return Newick tree, Newick tree with replaced tags, group tags, original tags, and randomly generated colours for each tag</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ab9a6cbb74e0df49dd0ed6151cc023e8eb98181f/src/shared/shared.jl#L150-L154">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.optimize_MG94_F3x4-Tuple{Any, Any, Any}" href="#CodonMolecularEvolution.optimize_MG94_F3x4-Tuple{Any, Any, Any}"><code>CodonMolecularEvolution.optimize_MG94_F3x4</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">optimize_MG94_F3x4(seqnames, seqs, tree; leaf_name_transform=x -&gt; x, genetic_code=MolecularEvolution.universal_code)</code></pre><p>Optimizes the MG94+F3x4 model on a tree, given a set of sequences and a tree. Returns the optimized tree, alpha, beta, nuc<em>matrix, F3x4, and eq</em>freqs. The leaf<em>name</em>transform kwarg can be used to transform the leaf names in the tree to match the seqnames.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ab9a6cbb74e0df49dd0ed6151cc023e8eb98181f/src/shared/shared.jl#L191-L196">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="difFUBAR/">difFUBAR »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.25 on <span class="colophon-date" title="Tuesday 11 June 2024 10:42">Tuesday 11 June 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
