<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>difFUBAR · CodonMolecularEvolution.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://MurrellGroup.github.io/CodonMolecularEvolution.jl/difFUBAR/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/favicon.ico" rel="icon" type="image/x-icon"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="CodonMolecularEvolution.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">CodonMolecularEvolution.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>difFUBAR</a><ul class="internal"><li><a class="tocitem" href="#Example"><span>Example</span></a></li><li><a class="tocitem" href="#Interface"><span>Interface</span></a></li></ul></li><li><a class="tocitem" href="../shared/">Shared tools</a></li><li><a class="tocitem" href="../dataprep/">Data preparation</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>difFUBAR</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>difFUBAR</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/main/docs/src/difFUBAR.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="difFUBAR"><a class="docs-heading-anchor" href="#difFUBAR">difFUBAR</a><a id="difFUBAR-1"></a><a class="docs-heading-anchor-permalink" href="#difFUBAR" title="Permalink"></a></h1><p>An interface to <a href="https://academic.oup.com/mbe/article/30/5/1196/998247">difFUBAR (awaiting the correct link)</a></p><h2 id="Example"><a class="docs-heading-anchor" href="#Example">Example</a><a id="Example-1"></a><a class="docs-heading-anchor-permalink" href="#Example" title="Permalink"></a></h2><p>Reads codon sequences from <a href="https://raw.githubusercontent.com/MurrellGroup/CodonMolecularEvolution.jl/main/test/data/Ace2_no_background/Ace2_tiny_test.fasta">this FASTA file</a>, and a tagged phylogeny from <a href="https://raw.githubusercontent.com/MurrellGroup/CodonMolecularEvolution.jl/main/test/data/Ace2_no_background/Ace2_no_background.nex">this NEXUS tree file</a>.</p><pre><code class="language-julia hljs">using MolecularEvolution, FASTX, CodonMolecularEvolution
analysis_name = &quot;output/Ace2&quot;
seqnames,seqs = read_fasta(&quot;data/Ace2_tiny_test.fasta&quot;);
treestring, tags, tag_colors = import_colored_figtree_nexus_as_tagged_tree(&quot;data/Ace2_no_background.nex&quot;)
df,results = difFUBAR(seqnames, seqs, treestring, tags, analysis_name);</code></pre><pre><code class="nohighlight hljs">Step 1: Initialization. If exports = true, tree showing the assignment of branches to groups/colors will be exported to: output/Ace2_tagged_input_tree.svg.
Step 2: Optimizing global codon model parameters.
0.0% 29.0% 58.0% 87.0% 
Step 4: Running Gibbs sampler to infer site categories.
Step 5: Tabulating and plotting. Detected sites:
Site 3 - P(ω1 &gt; ω2):0.0075; P(ω2 &gt; ω1):0.9805; P(ω1 &gt; 1):0.1205; P(ω2 &gt; 1):0.9675
Site 13 - P(ω1 &gt; ω2):0.0175; P(ω2 &gt; ω1):0.9605; P(ω1 &gt; 1):0.13; P(ω2 &gt; 1):0.9305

If exports = true, writing results for all sites to CSV: output/Ace2_posteriors.csv
Plotting alpha and omega distributions. If exports = true, saved as output/Ace2_violin_*.pdf</code></pre><p><em>Here&#39;s a selection of the output files:</em></p><p>output/Ace2_tagged_input_tree.svg</p><p><img src="../figures/Ace2_tagged_input_tree.svg" alt/></p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>difFUBAR allows for an optional, untagged, background branch set.</p></div></div><p>output/Ace2_site_omega_means.pdf</p><p><img src="../figures/Ace2_site_omega_means.png" alt/></p><p>output/Ace2_violin_all_params.pdf</p><p><img src="../figures/Ace2_violin_all_params.png" alt/></p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The output has a stochastic element to it due to the MCMC.</p></div></div><h2 id="Interface"><a class="docs-heading-anchor" href="#Interface">Interface</a><a id="Interface-1"></a><a class="docs-heading-anchor-permalink" href="#Interface" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.difFUBAR" href="#CodonMolecularEvolution.difFUBAR"><code>CodonMolecularEvolution.difFUBAR</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">difFUBAR(seqnames, seqs, treestring, tags, outpath; &lt;keyword arguments&gt;)</code></pre><p>Takes a tagged phylogeny and an alignment as input and performs difFUBAR analysis.</p><p><strong>Arguments</strong></p><ul><li><code>seqnames</code>: vector of untagged sequence names.</li><li><code>seqs</code>: vector of aligned sequences, corresponding to <code>seqnames</code>.</li><li><code>treestring</code>: a tagged newick tree string.</li><li><code>tags</code>: vector of tag signatures.</li><li><code>outpath</code>: export directory.</li><li><code>tag_colors=DIFFUBAR_TAG_COLORS[sortperm(tags)]</code>: vector of tag colors (hex format). The default option is consistent with the difFUBAR paper (Foreground 1: red, Foreground 2: blue).</li><li><code>pos_thresh=0.95</code>: threshold of significance for the posteriors.</li><li><code>iters=2500</code>: iterations used in the Gibbs sampler.</li><li><code>verbosity=1</code>: as verbosity increases, prints are added accumulatively. <ul><li>0 - no prints</li><li>1 - show current step and where output files are exported</li><li>2 - show the chosen <code>difFUBAR_grid</code> version and amount of parallel threads.</li></ul></li><li><code>exports=true</code>: if true, output files are exported.</li><li><code>code=MolecularEvolution.universal_code</code>: genetic code used for the analysis.</li><li><code>optimize_branch_lengths=false</code>: if true, the branch lengths of the phylogenetic tree are optimized.</li><li><code>version::Union{difFUBARGrid, Nothing}=nothing</code>: explicitly choose the version of <code>difFUBAR_grid</code> to use. If <code>nothing</code>, the version is heuristically chosen based on the available RAM and Julia threads.</li><li><code>t=0</code>: explicitly choose the amount of Julia threads to use. If <code>0</code>, the degree of parallelization is heuristically chosen based on the available RAM and Julia threads.</li></ul><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Julia starts up with a single thread of execution, by default. See <a href="https://docs.julialang.org/en/v1/manual/multi-threading/#Starting-Julia-with-multiple-threads">Starting Julia with multiple threads</a>.</p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/c6daed9f94b7603138c68ea6fc5cbe87803d6ff7/src/difFUBAR/difFUBAR.jl#L426-L452">source</a></section></article><h3 id="For-a-simple-and-often-optimal-configuration"><a class="docs-heading-anchor" href="#For-a-simple-and-often-optimal-configuration">For a simple and often optimal configuration</a><a id="For-a-simple-and-often-optimal-configuration-1"></a><a class="docs-heading-anchor-permalink" href="#For-a-simple-and-often-optimal-configuration" title="Permalink"></a></h3><ul><li>Launch Julia in the following manner: <code>julia -t auto</code></li><li>Keep the default values of the kwargs <code>version</code> and <code>t</code></li></ul><p>This lets Julia decide the amount of Julia threads and lets CodonMolecularEvolution.jl decide the <a href="#difFUBARGrid">difFUBARGrid</a> subtype to dispatch on and the degree of parallelization.</p><h3 id="difFUBARGrid"><a class="docs-heading-anchor" href="#difFUBARGrid">difFUBARGrid</a><a id="difFUBARGrid-1"></a><a class="docs-heading-anchor-permalink" href="#difFUBARGrid" title="Permalink"></a></h3><p><strong>Subtypes</strong> that decide which method to use for the grid likelihood computations.</p><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.difFUBARBaseline" href="#CodonMolecularEvolution.difFUBARBaseline"><code>CodonMolecularEvolution.difFUBARBaseline</code></a> — <span class="docstring-category">Type</span></header><section><div><p><strong>Constructor</strong></p><pre><code class="nohighlight hljs">difFUBARBaseline()</code></pre><p><strong>Description</strong></p><p>Use the trivial implementation of the grid likelihood computations, i.e. 1 thread without sub-tree likelihood caching.</p><p>See also: <code>difFUBARParallel</code>, <code>difFUBARTreesurgery</code>, <code>difFUBARTreesurgeryAndParallel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/c6daed9f94b7603138c68ea6fc5cbe87803d6ff7/src/difFUBAR/grids.jl#L7-L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.difFUBARParallel" href="#CodonMolecularEvolution.difFUBARParallel"><code>CodonMolecularEvolution.difFUBARParallel</code></a> — <span class="docstring-category">Type</span></header><section><div><p><strong>Constructor</strong></p><pre><code class="nohighlight hljs">difFUBARParallel()</code></pre><p><strong>Description</strong></p><p>Extend the baseline version by parallelizing the grid calculations. Requires julia to be launched with the <code>t</code> switch. Using <code>t</code> computational threads, where <code>t</code> is sufficiently small, memory complexity is usually O(t) and time complexity O(1/t). Empirical tests suggests that <code>t</code> should not be higher than the machine&#39;s total CPU threads and usually not higher than half of it&#39;s total threads.</p><p>See also: <code>difFUBARBaseline</code>, <code>difFUBARTreesurgery</code>, <code>difFUBARTreesurgeryAndParallel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/c6daed9f94b7603138c68ea6fc5cbe87803d6ff7/src/difFUBAR/grids.jl#L17-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.difFUBARTreesurgery" href="#CodonMolecularEvolution.difFUBARTreesurgery"><code>CodonMolecularEvolution.difFUBARTreesurgery</code></a> — <span class="docstring-category">Type</span></header><section><div><p><strong>Constructor</strong></p><pre><code class="nohighlight hljs">difFUBARTreesurgery()</code></pre><p><strong>Description</strong></p><p>Use sub-tree likelihood caching described in the &quot;Methods&quot; section of the difFUBAR paper. Use more memory than the baseline version but be significantly faster, if purity is high.</p><p>See also: <code>difFUBARBaseline</code>, <code>difFUBARParallel</code>, <code>difFUBARTreesurgeryAndParallel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/c6daed9f94b7603138c68ea6fc5cbe87803d6ff7/src/difFUBAR/grids.jl#L29-L39">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CodonMolecularEvolution.difFUBARTreesurgeryAndParallel" href="#CodonMolecularEvolution.difFUBARTreesurgeryAndParallel"><code>CodonMolecularEvolution.difFUBARTreesurgeryAndParallel</code></a> — <span class="docstring-category">Type</span></header><section><div><p><strong>Constructor</strong></p><pre><code class="nohighlight hljs">difFUBARTreesurgeryAndParallel()</code></pre><p><strong>Description</strong></p><p>Use parallelization and sub-tree likelihood caching. The most performant version in most cases. Use more memory than other versions.</p><p>See also: <code>difFUBARBaseline</code>, <code>difFUBARTreesurgery</code>, <code>difFUBARParallel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/c6daed9f94b7603138c68ea6fc5cbe87803d6ff7/src/difFUBAR/grids.jl#L41-L49">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../shared/">Shared tools »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.25 on <span class="colophon-date" title="Thursday 9 January 2025 16:09">Thursday 9 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
