<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>difFUBAR · CodonMolecularEvolution.jl</title><meta name="title" content="difFUBAR · CodonMolecularEvolution.jl"/><meta property="og:title" content="difFUBAR · CodonMolecularEvolution.jl"/><meta property="twitter:title" content="difFUBAR · CodonMolecularEvolution.jl"/><meta name="description" content="Documentation for CodonMolecularEvolution.jl."/><meta property="og:description" content="Documentation for CodonMolecularEvolution.jl."/><meta property="twitter:description" content="Documentation for CodonMolecularEvolution.jl."/><meta property="og:url" content="https://MurrellGroup.github.io/CodonMolecularEvolution.jl/difFUBAR/"/><meta property="twitter:url" content="https://MurrellGroup.github.io/CodonMolecularEvolution.jl/difFUBAR/"/><link rel="canonical" href="https://MurrellGroup.github.io/CodonMolecularEvolution.jl/difFUBAR/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/favicon.ico" rel="icon" type="image/x-icon"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="CodonMolecularEvolution.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">CodonMolecularEvolution.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>difFUBAR</a><ul class="internal"><li><a class="tocitem" href="#Example"><span>Example</span></a></li><li><a class="tocitem" href="#Interface"><span>Interface</span></a></li></ul></li><li><a class="tocitem" href="../shared/">Shared tools</a></li><li><a class="tocitem" href="../dataprep/">Data preparation</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>difFUBAR</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>difFUBAR</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/main/docs/src/difFUBAR.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="difFUBAR"><a class="docs-heading-anchor" href="#difFUBAR">difFUBAR</a><a id="difFUBAR-1"></a><a class="docs-heading-anchor-permalink" href="#difFUBAR" title="Permalink"></a></h1><p>An interface to <a href="https://academic.oup.com/mbe/article/30/5/1196/998247">difFUBAR (awaiting the correct link)</a></p><h2 id="Example"><a class="docs-heading-anchor" href="#Example">Example</a><a id="Example-1"></a><a class="docs-heading-anchor-permalink" href="#Example" title="Permalink"></a></h2><p>Reads codon sequences from <a href="https://raw.githubusercontent.com/MurrellGroup/CodonMolecularEvolution.jl/main/test/data/Ace2_tiny/Ace2_tiny_tagged.fasta">this FASTA file</a>, and a tagged phylogeny from <a href="https://raw.githubusercontent.com/MurrellGroup/CodonMolecularEvolution.jl/main/test/data/Ace2_tiny/Ace2_tiny_tagged.nex">this NEXUS tree file</a>.</p><pre><code class="language-julia hljs">using MolecularEvolution, FASTX, CodonMolecularEvolution
#Trigger PlotsExt
using Plots, Phylo
analysis_name = &quot;output/Ace2&quot;
seqnames,seqs = read_fasta(&quot;Ace2_tiny_tagged.fasta&quot;);
treestring, tags, tag_colors = import_colored_figtree_nexus_as_tagged_tree(&quot;Ace2_tiny_tagged.nex&quot;)
df,results = difFUBAR(seqnames, seqs, treestring, tags, analysis_name);</code></pre><pre><code class="nohighlight hljs">Step 1: Initialization. If exports = true, tree showing the assignment of branches to groups/colors will be exported to: output/Ace2_tiny_tagged_input_tree.svg.
Step 2: Optimizing global codon model parameters.
Optimized single α,β LL=-17461.802926840995 with α=1.6881756897953715 and β=0.8305228439241399.
Step 3: Calculating grid of 12096-by-808 conditional likelihood values (the slowest step). Currently on:
0.0% 4.0% 8.0% 12.0% 17.0% 21.0% 25.0% 29.0% 33.0% 37.0% 41.0% 45.0% 50.0% 54.0% 58.0% 62.0% 66.0% 70.0% 74.0% 79.0% 83.0% 87.0% 91.0% 95.0% 99.0% 
Step 4: Running Gibbs sampler to infer site categories.
Step 5: Tabulating and plotting. Detected sites:
Site 3 - P(ω1 &gt; ω2):0.0185; P(ω2 &gt; ω1):0.9655; P(ω1 &gt; 1):0.1285; P(ω2 &gt; 1):0.941
Site 26 - P(ω1 &gt; ω2):0.9765; P(ω2 &gt; ω1):0.007; P(ω1 &gt; 1):0.9395; P(ω2 &gt; 1):0.1205
Site 29 - P(ω1 &gt; ω2):0.9865; P(ω2 &gt; ω1):0.003; P(ω1 &gt; 1):0.9985; P(ω2 &gt; 1):0.1785
Site 36 - P(ω1 &gt; ω2):0.9965; P(ω2 &gt; ω1):0.0005; P(ω1 &gt; 1):0.951; P(ω2 &gt; 1):0.027
Site 68 - P(ω1 &gt; ω2):0.9665; P(ω2 &gt; ω1):0.0145; P(ω1 &gt; 1):0.568; P(ω2 &gt; 1):0.0255
Site 69 - P(ω1 &gt; ω2):0.2265; P(ω2 &gt; ω1):0.4945; P(ω1 &gt; 1):0.971; P(ω2 &gt; 1):0.977
Site 75 - P(ω1 &gt; ω2):0.0105; P(ω2 &gt; ω1):0.9735; P(ω1 &gt; 1):0.088; P(ω2 &gt; 1):0.915
Site 89 - P(ω1 &gt; ω2):0.946; P(ω2 &gt; ω1):0.0195; P(ω1 &gt; 1):0.978; P(ω2 &gt; 1):0.227
Site 93 - P(ω1 &gt; ω2):0.977; P(ω2 &gt; ω1):0.0065; P(ω1 &gt; 1):0.9405; P(ω2 &gt; 1):0.0635
Site 94 - P(ω1 &gt; ω2):0.9825; P(ω2 &gt; ω1):0.0035; P(ω1 &gt; 1):0.975; P(ω2 &gt; 1):0.1525
Site 101 - P(ω1 &gt; ω2):0.951; P(ω2 &gt; ω1):0.0175; P(ω1 &gt; 1):0.8605; P(ω2 &gt; 1):0.113
Site 214 - P(ω1 &gt; ω2):0.925; P(ω2 &gt; ω1):0.0175; P(ω1 &gt; 1):0.9975; P(ω2 &gt; 1):0.422
Site 215 - P(ω1 &gt; ω2):0.968; P(ω2 &gt; ω1):0.0095; P(ω1 &gt; 1):0.9025; P(ω2 &gt; 1):0.0915
Site 218 - P(ω1 &gt; ω2):0.98; P(ω2 &gt; ω1):0.0045; P(ω1 &gt; 1):0.969; P(ω2 &gt; 1):0.0925
Site 220 - P(ω1 &gt; ω2):0.003; P(ω2 &gt; ω1):0.9835; P(ω1 &gt; 1):0.2815; P(ω2 &gt; 1):0.986
Site 251 - P(ω1 &gt; ω2):0.0125; P(ω2 &gt; ω1):0.967; P(ω1 &gt; 1):0.0635; P(ω2 &gt; 1):0.7555
Site 282 - P(ω1 &gt; ω2):0.974; P(ω2 &gt; ω1):0.009; P(ω1 &gt; 1):0.9265; P(ω2 &gt; 1):0.099
Site 288 - P(ω1 &gt; ω2):0.93; P(ω2 &gt; ω1):0.0265; P(ω1 &gt; 1):0.957; P(ω2 &gt; 1):0.181
Site 361 - P(ω1 &gt; ω2):0.012; P(ω2 &gt; ω1):0.959; P(ω1 &gt; 1):0.528; P(ω2 &gt; 1):0.988
Site 366 - P(ω1 &gt; ω2):0.0135; P(ω2 &gt; ω1):0.972; P(ω1 &gt; 1):0.0945; P(ω2 &gt; 1):0.8965
Site 389 - P(ω1 &gt; ω2):0.9815; P(ω2 &gt; ω1):0.006; P(ω1 &gt; 1):0.9615; P(ω2 &gt; 1):0.0645
Site 422 - P(ω1 &gt; ω2):0.9675; P(ω2 &gt; ω1):0.0095; P(ω1 &gt; 1):0.9185; P(ω2 &gt; 1):0.084
Site 431 - P(ω1 &gt; ω2):0.988; P(ω2 &gt; ω1):0.0025; P(ω1 &gt; 1):0.994; P(ω2 &gt; 1):0.123
Site 533 - P(ω1 &gt; ω2):0.9535; P(ω2 &gt; ω1):0.021; P(ω1 &gt; 1):0.9235; P(ω2 &gt; 1):0.162
Site 557 - P(ω1 &gt; ω2):0.8025; P(ω2 &gt; ω1):0.0685; P(ω1 &gt; 1):0.974; P(ω2 &gt; 1):0.64
Site 570 - P(ω1 &gt; ω2):0.9805; P(ω2 &gt; ω1):0.0035; P(ω1 &gt; 1):0.97; P(ω2 &gt; 1):0.104
Site 574 - P(ω1 &gt; ω2):0.9805; P(ω2 &gt; ω1):0.0065; P(ω1 &gt; 1):0.9515; P(ω2 &gt; 1):0.0995
Site 580 - P(ω1 &gt; ω2):0.9655; P(ω2 &gt; ω1):0.009; P(ω1 &gt; 1):0.624; P(ω2 &gt; 1):0.04
Site 609 - P(ω1 &gt; ω2):0.9705; P(ω2 &gt; ω1):0.0095; P(ω1 &gt; 1):0.9605; P(ω2 &gt; 1):0.133
Site 661 - P(ω1 &gt; ω2):0.6195; P(ω2 &gt; ω1):0.161; P(ω1 &gt; 1):0.99; P(ω2 &gt; 1):0.9445
Site 697 - P(ω1 &gt; ω2):0.9515; P(ω2 &gt; ω1):0.02; P(ω1 &gt; 1):0.9665; P(ω2 &gt; 1):0.166
Site 709 - P(ω1 &gt; ω2):0.0035; P(ω2 &gt; ω1):0.986; P(ω1 &gt; 1):0.1365; P(ω2 &gt; 1):0.8755
Site 776 - P(ω1 &gt; ω2):0.0105; P(ω2 &gt; ω1):0.965; P(ω1 &gt; 1):0.1385; P(ω2 &gt; 1):0.915
Site 787 - P(ω1 &gt; ω2):0.02; P(ω2 &gt; ω1):0.942; P(ω1 &gt; 1):0.4145; P(ω2 &gt; 1):0.9505
Site 791 - P(ω1 &gt; ω2):0.0115; P(ω2 &gt; ω1):0.9645; P(ω1 &gt; 1):0.1155; P(ω2 &gt; 1):0.908

If exports = true, writing results for all sites to CSV: output/Ace2_tiny_posteriors.csv
Plotting alpha and omega distributions. If exports = true, saved as output/Ace2_tiny_violin_*.pdf</code></pre><p><em>Here&#39;s a selection of the output files:</em></p><p>output/Ace2_tiny_tagged_input_tree.svg</p><p><img src="../figures/output/Ace2_tiny_tagged_input_tree.svg" alt/></p><div class="admonition is-info" id="Note-b286c379052c0294"><header class="admonition-header">Note<a class="admonition-anchor" href="#Note-b286c379052c0294" title="Permalink"></a></header><div class="admonition-body"><p>difFUBAR allows for an optional, untagged, background branch set.</p></div></div><p>output/Ace2_tiny_site_omega_means.pdf</p><p><img src="../figures/output/Ace2_tiny_site_omega_means.png" alt/></p><p>output/Ace2_tiny_violin_all_params.pdf</p><p><img src="../figures/output/Ace2_tiny_violin_all_params.png" alt/></p><div class="admonition is-info" id="Note-de2431a052bcd7db"><header class="admonition-header">Note<a class="admonition-anchor" href="#Note-de2431a052bcd7db" title="Permalink"></a></header><div class="admonition-body"><p>The output has a stochastic element to it due to the MCMC.</p></div></div><h2 id="Interface"><a class="docs-heading-anchor" href="#Interface">Interface</a><a id="Interface-1"></a><a class="docs-heading-anchor-permalink" href="#Interface" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CodonMolecularEvolution.difFUBAR-difFUBAR" href="#CodonMolecularEvolution.difFUBAR-difFUBAR"><code>CodonMolecularEvolution.difFUBAR</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">difFUBAR(seqnames, seqs, treestring, tags, outpath; &lt;keyword arguments&gt;)</code></pre><p>Takes a tagged phylogeny and an alignment as input and performs difFUBAR analysis. Returns <code>df, results_tuple, plots_named_tuple</code> where <code>df</code> is a DataFrame of the detected sites, <code>results_tuple</code> is a tuple of the partial calculations needed to re-run <code>difFUBAR_tabulate_and_plot</code>, and <code>plots_named_tuple</code> is a named tuple of plots. Consistent with the docs of <a href="../#CodonMolecularEvolution.difFUBAR_tabulate_and_plot-NTuple{7, Any}"><code>difFUBAR_tabulate_and_plot</code></a>, <code>results_tuple</code> stores <code>(alloc_grid, codon_param_vec, alphagrid, omegagrid, tag_colors)</code>.</p><p><strong>Arguments</strong></p><ul><li><code>seqnames</code>: vector of untagged sequence names.</li><li><code>seqs</code>: vector of aligned sequences, corresponding to <code>seqnames</code>.</li><li><code>treestring</code>: a tagged newick tree string.</li><li><code>tags</code>: vector of tag signatures.</li><li><code>outpath</code>: export directory.</li><li><code>tag_colors=DIFFUBAR_TAG_COLORS[sortperm(tags)]</code>: vector of tag colors (hex format). The default option is consistent with the difFUBAR paper (Foreground 1: red, Foreground 2: blue).</li><li><code>pos_thresh=0.95</code>: threshold of significance for the posteriors.</li><li><code>iters=2500</code>: iterations used in the Gibbs sampler.</li><li><code>binarize=false</code>: if true, the tree is binarized before the analysis.</li><li><code>verbosity=1</code>: as verbosity increases, prints are added accumulatively. <ul><li>0 - no prints</li><li>1 - show current step and where output files are exported</li><li>2 - show the chosen <code>difFUBAR_grid</code> version and amount of parallel threads.</li></ul></li><li><code>exports=true</code>: if true, output files are exported.</li><li><code>code=MolecularEvolution.universal_code</code>: genetic code used for the analysis.</li><li><code>optimize_branch_lengths=false</code>: if true, the branch lengths of the phylogenetic tree are optimized.</li><li><code>version::Union{difFUBARGrid, Nothing}=nothing</code>: explicitly choose the version of <code>difFUBAR_grid</code> to use. If <code>nothing</code>, the version is heuristically chosen based on the available RAM and Julia threads.</li><li><code>t=0</code>: explicitly choose the amount of Julia threads to use. If <code>0</code>, the degree of parallelization is heuristically chosen based on the available RAM and Julia threads.</li></ul><div class="admonition is-info" id="Note-b03068c57dce8771"><header class="admonition-header">Note<a class="admonition-anchor" href="#Note-b03068c57dce8771" title="Permalink"></a></header><div class="admonition-body"><p>Julia starts up with a single thread of execution, by default. See <a href="https://docs.julialang.org/en/v1/manual/multi-threading/#Starting-Julia-with-multiple-threads">Starting Julia with multiple threads</a>.</p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ce5365ac6c0c2d680875eb578fb48b0454f0d92f/src/difFUBAR/difFUBAR.jl#L278-L307">source</a></section></article><h3 id="For-a-simple-and-often-optimal-configuration"><a class="docs-heading-anchor" href="#For-a-simple-and-often-optimal-configuration">For a simple and often optimal configuration</a><a id="For-a-simple-and-often-optimal-configuration-1"></a><a class="docs-heading-anchor-permalink" href="#For-a-simple-and-often-optimal-configuration" title="Permalink"></a></h3><ul><li>Launch Julia in the following manner: <code>julia -t auto</code></li><li>Keep the default values of the kwargs <code>version</code> and <code>t</code></li></ul><p>This lets Julia decide the amount of Julia threads and lets CodonMolecularEvolution.jl decide the <a href="#difFUBARGrid">difFUBARGrid</a> subtype to dispatch on and the degree of parallelization.</p><h3 id="difFUBARGrid"><a class="docs-heading-anchor" href="#difFUBARGrid">difFUBARGrid</a><a id="difFUBARGrid-1"></a><a class="docs-heading-anchor-permalink" href="#difFUBARGrid" title="Permalink"></a></h3><p><strong>Subtypes</strong> that decide which method to use for the grid likelihood computations.</p><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CodonMolecularEvolution.difFUBARBaseline-difFUBAR" href="#CodonMolecularEvolution.difFUBARBaseline-difFUBAR"><code>CodonMolecularEvolution.difFUBARBaseline</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Constructor</strong></p><pre><code class="nohighlight hljs">difFUBARBaseline()</code></pre><p><strong>Description</strong></p><p>Use the trivial implementation of the grid likelihood computations, i.e. 1 thread without sub-tree likelihood caching.</p><p>See also: <code>difFUBARParallel</code>, <code>difFUBARTreesurgery</code>, <code>difFUBARTreesurgeryAndParallel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ce5365ac6c0c2d680875eb578fb48b0454f0d92f/src/difFUBAR/grids.jl#L7-L15">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CodonMolecularEvolution.difFUBARParallel-difFUBAR" href="#CodonMolecularEvolution.difFUBARParallel-difFUBAR"><code>CodonMolecularEvolution.difFUBARParallel</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Constructor</strong></p><pre><code class="nohighlight hljs">difFUBARParallel()</code></pre><p><strong>Description</strong></p><p>Extend the baseline version by parallelizing the grid calculations. Requires julia to be launched with the <code>t</code> switch. Using <code>t</code> computational threads, where <code>t</code> is sufficiently small, memory complexity is usually O(t) and time complexity O(1/t). Empirical tests suggests that <code>t</code> should not be higher than the machine&#39;s total CPU threads and usually not higher than half of it&#39;s total threads.</p><p>See also: <code>difFUBARBaseline</code>, <code>difFUBARTreesurgery</code>, <code>difFUBARTreesurgeryAndParallel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ce5365ac6c0c2d680875eb578fb48b0454f0d92f/src/difFUBAR/grids.jl#L17-L27">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CodonMolecularEvolution.difFUBARTreesurgery-difFUBAR" href="#CodonMolecularEvolution.difFUBARTreesurgery-difFUBAR"><code>CodonMolecularEvolution.difFUBARTreesurgery</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Constructor</strong></p><pre><code class="nohighlight hljs">difFUBARTreesurgery()</code></pre><p><strong>Description</strong></p><p>Use sub-tree likelihood caching described in the &quot;Methods&quot; section of the difFUBAR paper. Use more memory than the baseline version but be significantly faster, if purity is high.</p><p>See also: <code>difFUBARBaseline</code>, <code>difFUBARParallel</code>, <code>difFUBARTreesurgeryAndParallel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ce5365ac6c0c2d680875eb578fb48b0454f0d92f/src/difFUBAR/grids.jl#L29-L39">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CodonMolecularEvolution.difFUBARTreesurgeryAndParallel-difFUBAR" href="#CodonMolecularEvolution.difFUBARTreesurgeryAndParallel-difFUBAR"><code>CodonMolecularEvolution.difFUBARTreesurgeryAndParallel</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Constructor</strong></p><pre><code class="nohighlight hljs">difFUBARTreesurgeryAndParallel()</code></pre><p><strong>Description</strong></p><p>Use parallelization and sub-tree likelihood caching. The most performant version in most cases. Use more memory than other versions.</p><p>See also: <code>difFUBARBaseline</code>, <code>difFUBARTreesurgery</code>, <code>difFUBARParallel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CodonMolecularEvolution.jl/blob/ce5365ac6c0c2d680875eb578fb48b0454f0d92f/src/difFUBAR/grids.jl#L41-L49">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../shared/">Shared tools »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.11.0 on <span class="colophon-date" title="Saturday 10 May 2025 23:45">Saturday 10 May 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
